### PR TITLE
Close plugin jar file on classloader close and after retrieving name for updating

### LINKED ISSUE #8879: Update Plugins feature bug
### Expected behavior

The Update Plugins feature should work properly as described in the link below.
https://docs.papermc.io/paper/updating#step-2-update-plugins

### Observed/Actual behavior

The Update Plugins feature is experiencing an error (since build `#405`).
https://mclo.gs/SLAagve

### Plugin and Datapack List

```
[19:30:38 INFO]: Server Plugins (1):
[19:30:38 INFO]: Paper Plugins:
[19:30:38 INFO]: Bukkit Plugins:
[19:30:38 INFO]:  - ProtocolLib
```
```
[19:30:53 INFO]: There are 2 data packs enabled: [vanilla (built-in)], [file/bukkit (world)]
[19:30:53 INFO]: There are no more data packs available
```

### Paper version

```
[19:31:10 INFO]: Checking version, please wait...
[19:31:12 INFO]: This server is running Paper version git-Paper-409 (MC: 1.19.3) (Implementing API version 1.19.3-R0.1-SNAPSHOT) (Git: 8427c88)
You are running the latest version
```

### Other

_No response_

### PR DESCRIPTION
**Pull request by emilyy-dev**
This should fix #8879 and similar reports on the Discord server, I can't test it because I'm not on Windows, so perhaps it would be good to add the build-pr-jar label and ask for someone to test the update feature on Windows.

---
Download the paperclip jar for this pull request: [paper-8902.zip](https://nightly.link/PaperMC/Paper/actions/artifacts/574775468.zip)

### CODE REVIEW

File: patches/server/0013-Paper-Plugins.patch [hunk 1]
```diff
@@ -949,10 +949,8 @@ index 0000000000000000000000000000000000000000..392eb260b44d5f9e685ce09596c19f0a
 +
 +    @Override
 +    public void close() throws IOException {
-+        try {
++        try (this.jar; this.libraryLoader) {
 +            super.close();
```
Comment (emilyy-dev) [inline patches/server/0013-Paper-Plugins.patch:1008]: Yes I am aware this looks odd, but doing a try-with-resources (which can also take local vars and fields as-is) will ensure that a) everything is closed, so I don't have to cascade 2 or 3 different `try-finally`s, and b) it also handles cases where IOExceptions are thrown, it doesn't discard anything by "appending" exceptions via `Throwable::addSuppressed` and rethrows it

**Comment by emilyy-dev** [comment]
Also, something worth noting because I asked myself the same thing so it should probably be made public to avoid confusion:

**Why am I not closing the JarFile in the PaperSimplePluginClassLoader instead, if the field is declared there?**

PSPCL's constructor is not only called from the PaperPluginClassLoader (which is the file this patches) through `super`, but it's used "standalone" for the PluginLoader/classpath builder mechanic, the JarFile given in that case is later on reused and passed to the PaperPluginClassLoader, right after the Simple has been used.

**Comment by Owen1212055** [comment]
This may cause issues with /reloading until https://github.com/PaperMC/Paper/pull/8881 is resolved.
This is because the underlying jar providers will be closed, which the providers still will be using.

**Comment by emilyy-dev** [comment]
Good catch, although I do think it is only sensible to close the JarFile when we're done using it, *especially* in an environment where hot-reloading is (semi-)supported. Will wait until that issue is resolved and continue from there.

**Comment by emilyy-dev** [comment]
I think I will repurpose this PR as two separate ones instead, one for closing the jar in getPluginName for the updating mechanic, and maybe another one for closing it in the classloader, but that would be after that other PR.
For the first issue, should I close this PR and open a new one or just shrink this one?

**Comment by Owen1212055** [comment]
> I think I will repurpose this PR as two separate ones instead, one for closing the jar in getPluginName for the updating mechanic, and maybe another one for closing it in the classloader, but that would be after that other PR.
> For the first issue, should I close this PR and open a new one or just shrink this one?

The required PR in question will be merged soon. It’s better to just wait a bit here.

**Comment by emilyy-dev** [comment]
Got to test this on a Windows computer, can confirm the update system works proper on startup for all plugins, and on reload for Paper plugins, Bukkit/Spigot plugins will fail to update on reload on Windows as the jar file is left untouched (as it always has..)

**Review by kennytv — APPROVED** [review]

**Review by Owen1212055 — APPROVED** [review]

**Comment by radiant-ai** [comment]
This completely fixes `update` folder functionality on my Windows. Thank you!

**Pull request merged**